// zerodev-sdk/src/constants.ts
// For local testing
//
// export const BACKEND_URL = 'http://localhost:4000'
// export const PAYMASTER_URL = 'http://localhost:4001'
// export const BUNDLER_URL = 'http://localhost:5050/rpc'
// export const PAYMASTER_ABI = require('./abi/paymaster_abi.json')
// export const ENTRYPOINT_ADDRESS = '0xaec9CE12Edd21BD042967008d24fD4ab12C745E5'
// export const PAYMASTER_ADDRESS = '0x057ef64E23666F000b34aE31332854aCBd1c8544'
// export const WALLET_FACTORY_ADDRESS = '0x39Cc3B8632d89E68D5986Ff69D3e36127994C0C9'

export const BACKEND_URL = process.env.REACT_APP_ZERODEV_STAGING ? 'https://staging-backend-51hj.onrender.com' : 'https://backend-vikp.onrender.com'
export const PAYMASTER_URL = process.env.REACT_APP_ZERODEV_STAGING ? 'https://staging-paymaster.onrender.com' : 'https://paymaster-server.onrender.com'
export const PAYMASTER_ABI = require('./abi/paymaster_abi.json')
export const ENTRYPOINT_ADDRESS: { [key: string]: any } = {
  '5': '0x49bd26a61DABE4D9F5B8e2B4fD947B9b250D30B9',
  '137': '0xaec9CE12Edd21BD042967008d24fD4ab12C745E5',
  '43113': '0x49bd26a61DABE4D9F5B8e2B4fD947B9b250D30B9',
  '43114': '0x49bd26a61DABE4D9F5B8e2B4fD947B9b250D30B9',
  '80001': '0xaec9CE12Edd21BD042967008d24fD4ab12C745E5',
}
export const WALLET_FACTORY_ADDRESS: { [key: string]: any } = {
  '5': '0x6D261eD72837f4001d4Cc89f9002ddb5AB3a93B2',
  '137': '0x6D261eD72837f4001d4Cc89f9002ddb5AB3a93B2',
  '43113': '0x6D261eD72837f4001d4Cc89f9002ddb5AB3a93B2',
  '43114': '0x6D261eD72837f4001d4Cc89f9002ddb5AB3a93B2',
  '80001': '0x6D261eD72837f4001d4Cc89f9002ddb5AB3a93B2',
}
export const BUNDLER_URL: { [key: string]: any } = {
  '5': 'https://goerli-bundler.onrender.com/rpc',
  '137': 'https://polygon-bundler.onrender.com/rpc',
  '43113': 'https://fuji-bundler.onrender.com/rpc',
  '43114': 'https://avalanche-bundler.onrender.com/rpc',
  '80001': 'https://bundler.onrender.com/rpc',
}
export const PAYMASTER_ADDRESS: { [key: string]: any } = {
  '5': '0x2E59ebb954BF7B0e1FB63c3B2F583C35C31Cd604',
  '137': '0x2E59ebb954BF7B0e1FB63c3B2F583C35C31Cd604',
  '43113': '0x2E59ebb954BF7B0e1FB63c3B2F583C35C31Cd604',
  '43114': '0x2E59ebb954BF7B0e1FB63c3B2F583C35C31Cd604',
  '80001': '0x73a073E6e3C2A0020995Fcc380ef86718c6ff6f3',
}

export const INFURA_API_KEY = 'f36f7f706a58477884ce6fe89165666c'
export const CHAIN_ID_TO_INFURA_NAMES: { [key: string]: any } = {
  '1': 'mainnet',
  '5': 'goerli',
  '137': 'polygon-mainnet',
  '80001': 'polygon-mumbai',
  '10': 'optimism-mainnet',
  '420': 'optimism-goerli',
  '42161': 'arbitrum-mainnet',
  '421613': 'arbitrum-goerli',
  '43114': 'avalanche-mainnet',
  '43113': 'avalanche-fuji',
  '1313161554': 'aurora-mainnet',
  '1313161555': 'aurora-testnet',
}
